File: 2106-maximum-fruits-harvested-after-at-most-k-steps/2106-maximum-fruits-harvested-after-at-most-k-steps.py
class Solution:
    def maxTotalFruits(self, fruits: List[List[int]], startPos: int, k: int) -> int:
        j = s = ans = 0
        for i, [pos, amount] in enumerate(fruits):
            s += amount
            while j <= i and min(abs(startPos - pos), abs(startPos - fruits[j][0])) + pos - fruits[j][0] > k:
                s -= fruits[j][1]
                j += 1
            ans = max(ans, s)
        return ans

File: 2106-maximum-fruits-harvested-after-at-most-k-steps/2106-maximum-fruits-harvested-after-at-most-k-steps.js
/**
 * @param {number[][]} fruits
 * @param {number} startPos
 * @param {number} k
 * @return {number}
 */
var maxTotalFruits = function(fruits, startPos, k) {
  const n = fruits.length;
  let [j, sum, ans] = [0, 0, 0];
  
  for (let [i, [pos, amount]] of fruits.entries()) {
    sum += amount;
    while (j <= i && pos - fruits[j][0] + Math.min(Math.abs(pos - startPos), Math.abs(startPos - fruits[j][0])) > k)
      sum -= fruits[j++][1];
    
    ans = Math.max(ans, sum);
  }
  
  return ans;
};